### ISSUE TITLE
Being able to rescue from OAuth2::Error

### ISSUE THREAD

**Issue by Dorian** [mkdynamic/omniauth-facebook#170]
Hello,

For some reason I get this error 

> OAuth2::Error#
> : {"error":{"message":"This authorization code has been used.","type":"OAuthException","code":100}}

But I can't find a way to rescue it (because it's in the before_action callbacks) and not let it bubble up to the user.

Any thought on it? How could I rescue it? Is it an OAuth2 issue? An Omniauth issue? An ominauth-facebook issue?

![screen shot 2014-11-11 at 2 25 02 pm](https://cloud.githubusercontent.com/assets/129149/5000180/d35909da-69ae-11e4-8b89-837578a4c0db.png)

**Comment by jessieay** [comment]
+1 We have a similar issue and we think it is happening because the callback is being fired twice. Not sure why it is being fire twice, but I would look in the logs and see if that is the same case for you, @Dorian !

**Comment by Dorian** [comment]
ok, my fix was to bypass the before filters by rendering before executing them (it's a giant hack):

```
  before_filter :set_facebook_user

  def facebook
    fail 'This should never be executed'
  end

  def set_facebook_user
    @facebook_user = get_facebook_user || session["devise.facebook_user"]
    session["devise.facebook_user"] = @facebook_user if @facebook_user
    auth = FacebookAuthenticator.new(@facebook_user, params, current_user)

    if auth.logging_in?
      set_current_user(auth.user)
    end

    render json: auth.response_hash, status: auth.status
  end

  def get_facebook_user
    token = request.env["omniauth.auth"].try(:credentials).try(:token)
    return unless token
    Koala::Facebook::API.new(token).get_object('me')
  end
```

eg, the most important thing is the token (`request.env["omniauth.auth"].try(:credentials).try(:token)`) that I then manually use to fetch the user (and also it can't be reused so I save the user in a session).

**Comment by jessieay** [comment]
Update:

Our situation might have been totally different, but we realized that our error was that we were calling the request from a non-HTTPS URL and using an HTTPS URL in the callback. So, the two weren't matching. The logs did not make this clear. Since we saw two callbacks we thought that was the issue, but it might be that we were just reading the logs wrong. 

This issue only became apparent when we moved to client-side Facebook oauth, where we saw this javascript failure:

```
 XMLHttpRequest cannot load 'https://example.com. No 'Access-Control-Allow-Origin' header is present on the requested resource. Origin 'http://example.com' is therefore not allowed access. 
```

Once we forced SSL on our site so that the origin was always SSL, it was fixed (in both client-side oauth and non client-side oauth implementations).
